PHP API for registering blocks with Block Lab

lukecarbis opened the pull request:
I've implemented the following functions, which enable to ability to register blocks programatically:

`block_lab_add_block( $block_name, $block_config )`

and

`block_lab_add_field( $block_name, $field_name, $field_config )`

Both are well documented in their respective doc blocks, if you'd like to learn more about these functions and how to use them.

They need to run during the `block_lab_add_blocks` action hook.

Here's some example code I have been testing with:

```PHP
// One huge long array with lots defined.
block_lab_add_block(
	'one-fish',
	array(
		'title'    => 'One Fish',
		'category' => 'common',
		'icon'     => 'waves',
		'excluded' => array( 'page' ),
		'keywords' => array( 'sad', 'glad', 'bad' ),
		'fields'   => array(
			'thin' => array(
				'label'   => 'Thin',
				'control' => 'toggle',
				'width'   => '25',
				'default' => true,
			),
			'fat' => array(
				'label'   => 'Fat',
				'control' => 'toggle',
				'width'   => '25',
				'default' => false,
			),
			'hat'  => array(
				'label'   => 'Hat',
				'control' => 'select',
				'width'   => '50',
				'options' => array(
					array(
						'label' => 'Yellow',
						'value' => 'yellow',
					),
					array(
						'label' => 'Red',
						'value' => 'red',
					),
					array(
						'label' => 'Blue',
						'value' => 'blue',
					),
				),
			),
		),
	)
);

// A more minimalist approach.
block_lab_add_block( 'two-fish', array( 'icon' => 'waves' ) );
block_lab_add_field( 'two-fish', 'hello-there-ned-how-do-you-do' );
block_lab_add_field( 'two-fish', 'tell-me-tell-me-what-is-new', array( 'placeholder' => 'My hat is old, my teeth are gold.' ) );
```

Closes #400.

lukecarbis committed:
Add block_lab_add_block helper function

lukecarbis committed:
Add block_lab_add_field helper function

lukecarbis committed:
Remove test code

lukecarbis committed:
Use more descriptive argument names in block registration helper functions

kienstra commented:
Hi @lukecarbis,
Nice, this will be really useful.

If it's alright, tomorrow is probably the earliest I could review this.

kienstra commented on php/helpers.php:
```diff
@@ -309,3 +309,136 @@ function block_field_config( $name ) {
+		$block_config['keywords'] = array();
+	}
+
+	if ( ! isset( $block_config['fields'] ) ) {
```
[wp_parse_args()](https://codex.wordpress.org/Function_Reference/wp_parse_args) might make this a little easier

lukecarbis replied:
Address in https://github.com/getblocklab/block-lab/pull/434/commits/e547cc32c00f439325e53d68d840695dbce54f2e.

kienstra commented on php/helpers.php:
```diff
@@ -309,3 +309,136 @@ function block_field_config( $name ) {
+		$block_config['fields'] = array();
+	}
+
+	add_filter(
```
The logic in the callback looks good.

What do you think about adding a public function to class `Loader` like:

`block_lab()->loader->add_block( $block_config );`

Then, it could store the registered blocks in a protected property of `Loader`.

That could have most of the logic in this callback, but this particular `add_filter()` call wouldn't be needed.

The `apply_filters( 'block_lab_blocks' )` call in `Loader` would still be useful, as would this `block_lab_add_block()` function.

lukecarbis replied:
Addressed in https://github.com/getblocklab/block-lab/pull/434/commits/5b98f71fd97a3e6fb6973b89c98f15bbb1993e10.

kienstra commented on php/helpers.php:
```diff
@@ -309,3 +309,136 @@ function block_field_config( $name ) {
+		$field_config['settings'] = array();
+	}
+
+	add_filter(
```
Like the comment above, what do you think about turning this `add_filter()` call into something like:

```php
block_lab()->loader->add_field( $block_name, $field_config );
```
...or whatever method name.

It could then add fields to blocks added via the [possible function above](https://github.com/getblocklab/block-lab/pull/434#discussion_r329407225): `block_lab()->loader->add_block( $block_config );`

kienstra replied:
Ah, that wouldn't be able to add fields to existing blocks created with `/wp-admin` > Block Lab > Add New.

Would that be a common use case, to add a block via via  `/wp-admin` > Block Lab > Add New, then later add a field to it with `block_lab_add_field()` ?

lukecarbis replied:
I don't think it would be a very common use case, no.

lukecarbis replied:
Addressed in https://github.com/getblocklab/block-lab/pull/434/commits/5b98f71fd97a3e6fb6973b89c98f15bbb1993e10.

kienstra replied:
Nice, that looks good. Thanks for pushing it.

kienstra reviewed (commented):
**Looks Good, Question About Adding Method To Loader**

Hi @lukecarbis,
This looks good, it's a big win for the plugin.

What do you think about moving the logic from the `add_filter()` calls into a new method in the `Loader` class?

This and #435 depend on the `Loader` class, but before, it wasn't really available in any other file.

#440 moves it into `block_lab()->loader`, and makes most of the method protected.

Either way, hope you're doing great, and look forward to talking tomorrow.

lukecarbis committed:
Merge branch 'develop' into feature/php-api

lukecarbis committed:
Use wp_parse_args for default block and field config

lukecarbis committed:
Store an array of blocks in the loader class, instead of a string of JSON

lukecarbis committed:
Merge branch 'develop' into feature/php-api

lukecarbis committed:
Move field type default to the field class

lukecarbis committed:
Change action name to add_block instead of register_block

lukecarbis committed:
Move block registration logic to the Loader

lukecarbis commented:
@kienstra Thanks for you're review! Great to have your eyes on it, great suggestions.

I've made some changes to address your comments. Could you please take another look?

kienstra commented:
Hi @lukecarbis, this looks really good. I think the logic is good to go 😄 

Would you mind if I pushed some unit tests to this branch just for completeness? For example, I was thinking of suggesting we eventually change to `array()` to the `[]` syntax, and it'd be good to have as much coverage as possible if we ever make that change.

lukecarbis commented:
Please feel free to push unit tests.

kienstra commented:
Thanks, Luke!

kienstra committed:
Add unit tests for add_field() and add_block()

Add assertions that these methods work.
Also, make a slight change to add_field().

kienstra committed:
Merge in develop, resolve conflicts

The conflicts were trivial.
For test-class-loader.php, I retained both edits.
It class-abstract-template.php, I used the
edits from the develop branch.
They were only to the @param tag in a DocBlock.

kienstra committed:
Add a few assertions for Field::from_array()

Test a case where there is not 'type' passed,
where it should add the type if the control exists.

kienstra committed:
Add a short test for block_lab_add_block() to see how Mockery works in Travis

This might not work well in Travis,
depending on how the version work together,
like the phpunit version.

kienstra committed:
Add another assertion for block_lab_add_block()

This tests using the second argument,
with a long name.

kienstra committed:
Add a test for block_lab_add_field()

This is similar to block_lab_add_block(),
and uses Mockery.

kienstra commented on composer.json:
```diff
@@ -9,6 +9,7 @@
     },
     "require-dev": {
         "brain/monkey": "^2",
+        "mockery/mockery": "^1.2.4",
```
This allows mocking entire classes, like `Loader`. It doesn't look like [Brain Monkey](https://brain-wp.github.io/BrainMonkey/docs/wordpress-hooks-added.html) can mock classes.

kienstra commented:
Using the PHP API [example above](https://github.com/getblocklab/block-lab/pull/434#issue-320112300), the block:

1. Looks as expected in the Edit component
2. Saves its values as expected
3. Looks as expected in the `<ServerSideRender>` component and on the front end

<img width="830" alt="expected-block-good" src="https://user-images.githubusercontent.com/4063887/67152654-bb258c00-f2a0-11e9-91c8-92136a4878bf.png">

kienstra reviewed (approved):
**Great work!**

Hi @lukecarbis,
This looks really good.

The API worked as expected, and I think it's ready to merge when you're ready.

kienstra commented on php/blocks/class-loader.php:
```diff
@@ -569,15 +569,15 @@ public function add_block( $block_config ) {
 		}
 		if ( ! isset( $field_config['name'] ) ) {
-			return $blocks;
+			return;
```
Is it alright that I made this change? Maybe something else was intended.

lukecarbis replied:
No, nothing else intended. Good pickup.
